src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SpeculativeTCompletionProviderTests.cs

```diff
@@ -162,6 +162,77 @@ class C
             await VerifyItemExistsAsync(markup, "T");
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37224, "https://github.com/dotnet/roslyn/issues/37224")]
+        public async Task InRef()
+        {
+            var markup = @"
+using System;
+class C
+{
+    ref $$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37224, "https://github.com/dotnet/roslyn/issues/37224")]
+        public async Task InRefGeneric()
+        {
+            var markup = @"
+using System;
+class C
+{
+    ref Func<$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37224, "https://github.com/dotnet/roslyn/issues/37224")]
+        public async Task InRefReadonlyGeneric()
+        {
+            var markup = @"
+using System;
+class C
+{
+    ref readonly Func<$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37268, "https://github.com/dotnet/roslyn/issues/37268")]
+        public async Task InQualifiedGeneric()
+        {
+            var markup = @"
+using System;
+class C
+{
+    System.Func<$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37224, "https://github.com/dotnet/roslyn/issues/37224")]
+        [WorkItem(37268, "https://github.com/dotnet/roslyn/issues/37268")]
+        public async Task InRefAndQualifiedGeneric()
+        {
+            var markup = @"
+using System;
+class C
+{
+    ref System.Func<$$
```

Comment:
can you add tests with:
1. nested generics
2. being after a comma in top-level and nested generics.

Reply:
We have cases for 1 and 2 right below, without `ref`, but I'm happy to add cases with `ref`.

Tuples! The current code does not support them. Congratulations we now have another bug #37361. I feel supporting tuples introduces more complex cases, let's address them separately.